### PR TITLE
Enable use of user-provided thread::Pool instance

### PR DESCRIPTION
**Pull request by mloskot**
Enable use of user-provided `thread::Pool` instance

`Pool` instance can be configured with user-specified number of threads and passed to `Reader` and `Writer` as optional argument of constructors.
If no `Pool` is specified, default instance is used and configured with number of threads based on actual number of cores on the system.

Advantages:
* Allows to use `Pool` configured with user-preferred number of threads.
* If libosmium used in a shared library, this helps to prevent potential deadlock during Pool shutdown at (after) shared library unloading (see #213 for details).

------
* Compiled with GCC 7.1.0 and passed all tests on Ubuntu Linux (complete test suite)
* Compiled with VS2017 and run using custom sample reader/writer (XML, PBF) app

### CODE REVIEW

File: test/t/thread/test_pool.cpp [hunk 1]
```diff
@@ -70,3 +70,30 @@ TEST_CASE("thread") {
...
+    SECTION("can send job to thread pool") {
+        auto future = pool.submit(test_job_with_result {});
+
+        REQUIRE(future.get() == 42);
+    }
+
+    SECTION("can throw from job in thread pool") {
+        auto future = pool.submit(test_job_throw {});
+
+        REQUIRE_THROWS_AS(future.get(), std::runtime_error);
```
Comment (joto) [inline test/t/thread/test_pool.cpp]: Should be `std::runtime_error&`.

**Comment by joto** [comment]
I think to be consistent, we should also do the same for `Reader`. But first get it right for the `Writer`.

File: include/osmium/io/writer.hpp [hunk 2]
```diff
@@ -236,9 +243,12 @@ namespace osmium {
                     (set_option(options, args), 0)...
                 };
 
-                if (options.header.get("generator") == "") {
-                    options.header.set("generator", "libosmium/" LIBOSMIUM_VERSION_STRING);
-                }
```
Comment (mloskot) [inline include/osmium/io/writer.hpp]: I was preparing the patch using 2.12.2, then diff-ported to master, loosing some post-2.12.2 changes, apparently. I'll restore it.

File: include/osmium/io/writer.hpp [hunk 3]
```diff
@@ -236,9 +243,12 @@ namespace osmium {
...
+                if (options.pool)
+                    m_pool = options.pool;
+                else
+                    m_pool = &thread::Pool::default_instance();
+
+                m_output = osmium::io::detail::OutputFormatFactory::instance().create_output(*m_pool, m_file, m_output_queue);
```
Comment (mloskot) [inline include/osmium/io/writer.hpp]: Your concern seems similar to the situation discussed here:
https://akrzemi1.wordpress.com/2012/11/14/not-using-stdthread/#comment-659

Isn't `Writer::m_thread` considered to actually be *"launching of the thread is the last instruction in the constructor body"* (if we shuffle `ensure_cleanup` a bit) ?

File: include/osmium/thread/pool.hpp [hunk 4]
```diff
@@ -162,7 +162,11 @@ namespace osmium {
 
             static constexpr int default_num_threads = 0;
 
-            static Pool& instance() {
+            static std::unique_ptr<Pool> create_instance(int num_threads = default_num_threads) {
+                return std::unique_ptr<Pool>(new Pool(num_threads, detail::get_work_queue_size()));
+            }
```
Comment (mloskot) [inline include/osmium/thread/pool.hpp]: > Do we need this function?

No, unless `Pool` ctor must be non-public, what I assumed.

I also aimed for some interface symmetry, more explicit and self-describing that one is supposed to use either `default_instance`  or `create_instance`, not mixing them.

If you prefer, I can remove it and make `Pool` ctor public. Let me know.

File: include/osmium/thread/pool.hpp [hunk 5]
```diff
@@ -178,6 +182,10 @@ namespace osmium {
                 shutdown_all_workers();
             }
 
+            int num_threads() const {
+                return m_num_threads;
```
Comment (mloskot) [inline include/osmium/thread/pool.hpp]: BTW, `queue_size`  and `queue_empty` are not specified with `noexcept`. Is it intentional or the refactoring is in progress?

File: include/osmium/io/detail/pbf_input_format.hpp [hunk 6]
```diff
@@ -183,7 +183,7 @@ namespace osmium {
                         PBFDataBlobDecoder data_blob_parser{std::move(input_buffer), read_types(), read_metadata()};
 
                         if (osmium::config::use_pool_threads_for_pbf_parsing()) {
-                            send_to_output_queue(osmium::thread::Pool::instance().submit(std::move(data_blob_parser)));
+                            send_to_output_queue(osmium::thread::Pool::default_instance().submit(std::move(data_blob_parser)));
```
Comment (mloskot) [inline include/osmium/io/detail/pbf_input_format.hpp]: Hmm, I'm trying to look closer and I can't see anything broken, comparing to the file in the current master.

pbf_input_format hpp

File: include/osmium/io/detail/output_format.hpp [hunk 7]
```diff
@@ -129,6 +132,7 @@ namespace osmium {
                 OutputFormat& operator=(const OutputFormat&) = delete;
                 OutputFormat& operator=(OutputFormat&&) = delete;
 
+
```
Comment (mloskot) [inline include/osmium/io/detail/output_format.hpp]: Unintentional. I'll fix it.

File: include/osmium/thread/pool.hpp (hunk 4 above)
Comment (joto) [inline include/osmium/thread/pool.hpp]: There is really no reason why the constructor can't be public. So please change that.

File: test/t/io/test_writer.cpp [hunk 8]
```diff
@@ -111,3 +111,56 @@ TEST_CASE("Writer") {
...
+    osmium::memory::Buffer buffer = reader.read();
+    REQUIRE(buffer);
+    REQUIRE(buffer.committed() > 0);
+    const auto num = std::distance(buffer.select<osmium::OSMObject>().cbegin(), buffer.select<osmium::OSMObject>().cend());
+    REQUIRE(num > 0);
+    REQUIRE(buffer.select<osmium::OSMObject>().cbegin()->id() == 1);
+
+    std::string filename;
+
+    SECTION("Successfull writes") {
```
Comment (mloskot) [inline test/t/io/test_writer.cpp]: A perfect copy-n-paste mistake :)

Fixed the first one in #215

File: test/t/io/test_writer.cpp [hunk 9]
```diff
@@ -111,3 +111,56 @@ TEST_CASE("Writer") {
...
+
+        SECTION("Writer buffer") {
```
Comment (mloskot) [inline test/t/io/test_writer.cpp]: Copying the original test was a quick hack.

No problem, I will refactor once we agree what to do about `create_instance` vs ctor.

File: test/t/io/test_writer.cpp [hunk 10]
```diff
@@ -111,3 +111,56 @@ TEST_CASE("Writer") {
...
+        }
+
+        osmium::io::Reader reader_check{filename};
+        osmium::memory::Buffer buffer_check = reader_check.read();
+        REQUIRE(buffer_check);
+        REQUIRE(buffer_check.committed() > 0);
+        REQUIRE(buffer_check.select<osmium::OSMObject>().size() == num);
+        REQUIRE(buffer_check.select<osmium::OSMObject>().cbegin()->id() == 1);
+    }
+}
```
Comment (mloskot) [inline test/t/io/test_writer.cpp]: Interesting. I suspect if this is not caused by git, due to [config.autocrlf=true](https://help.github.com/articles/dealing-with-line-endings/#global-settings-for-line-endings).

I don't use any auto-trimming in editor.

File: test/t/thread/test_pool.cpp [hunk 11]
```diff
@@ -70,3 +70,30 @@ TEST_CASE("thread") {
...
+    SECTION("can get access to thread pool") {
+        REQUIRE(pool.queue_empty());
+    }
+    
+    SECTION("can access user-provided number of threads") {
+        REQUIRE(pool.num_threads() == 7);
+    }
+
+    SECTION("can send job to thread pool") {
+        auto future = pool.submit(test_job_with_result {});
```
Comment (mloskot) [inline test/t/thread/test_pool.cpp]: The formatting is not mine, [it was there](https://github.com/osmcode/libosmium/blob/6a17a8ddd58c6d574259ac839fb550540e017eee/test/t/thread/test_pool.cpp#L60).

File: test/t/thread/test_pool.cpp (hunk 1 above)
Comment (mloskot) [inline test/t/thread/test_pool.cpp]: Same here, I copied [what was there](https://github.com/osmcode/libosmium/blob/6a17a8ddd58c6d574259ac839fb550540e017eee/test/t/thread/test_pool.cpp#L68), didn't I?

**Comment by mloskot** [comment]
> I think to be consistent, we should also do the same for Reader.

Yes.

> But first get it right for the Writer.

I like this approach.

File: test/t/thread/test_pool.cpp (hunk 1 above)
Comment (joto) [inline test/t/thread/test_pool.cpp]: Sorry, not your fault. My code was wrong. I'll fix my code. :-)

Comment (mloskot) [inline test/t/thread/test_pool.cpp]: No problem, cheers :-)

File: test/t/io/test_writer.cpp (hunk 10 above)
Comment (mloskot) [inline test/t/io/test_writer.cpp]: Should be fixed in the latest update 1d051f909b5d4a0aa5c76191e27300ff998398cc

File: include/osmium/thread/pool.hpp (hunk 4 above)
Comment (mloskot) [inline include/osmium/thread/pool.hpp]: I've removed `create_instance`  in 1d051f909b5d4a0aa5c76191e27300ff998398cc

File: test/t/thread/test_pool.cpp (hunk 11 above)
Comment (mloskot) [inline test/t/thread/test_pool.cpp]: I also fixed mine.

File: test/t/thread/test_pool.cpp (hunk 1 above)
Comment (mloskot) [inline test/t/thread/test_pool.cpp]: I also fixed mine.

File: test/t/io/test_writer.cpp (hunk 10 above)
Comment (mloskot) [inline test/t/io/test_writer.cpp]: Here is the current update and I have no idea why GitHub is not picking up the change (added new line) and not marking this review comment as outdated.

https://github.com/mloskot/libosmium/blob/a113611a206908823ca8931bfbd8e0bdf346372a/test/t/io/test_writer.cpp#L167

**Comment by mloskot** [comment]
@joto I think I've completed the first major round of review. If you could have a look at the updates and see what else needs improving.

File: include/osmium/thread/pool.hpp [hunk 12]
```diff
@@ -162,7 +162,11 @@ namespace osmium {
 
             static constexpr int default_num_threads = 0;
 
-            static Pool& instance() {
+            explicit Pool(int num_threads = default_num_threads) :
```
Comment (joto) [inline include/osmium/thread/pool.hpp]: Having a forwarding constructor here doesn't buy us anything. Just move the existing constructor down here so it becomes public and make the second parameter default to `detail::get_work_queue_size()`.

File: include/osmium/thread/pool.hpp [hunk 13]
```diff
@@ -162,7 +162,11 @@ namespace osmium {
...
+                Pool(num_threads, detail::get_work_queue_size()) {
+            }
+
+            static Pool& default_instance() {
                 static Pool pool(default_num_threads, detail::get_work_queue_size());
```
Comment (joto) [inline include/osmium/thread/pool.hpp]: This then becomes just `static Pool pool{}`.

File: test/t/io/test_writer.cpp [hunk 14]
```diff
@@ -111,3 +111,43 @@ TEST_CASE("Writer") {
...
+TEST_CASE("Writer with user-provided Pool") {
+    osmium::io::Header header;
+    header.set("generator", "test_writer.cpp");
+
+    osmium::io::Reader reader{with_data_dir("t/io/data.osm")};
+    osmium::memory::Buffer buffer = reader.read();
+    REQUIRE(buffer);
+    REQUIRE(buffer.committed() > 0);
+
+    SECTION("Successful writes") {
```
Comment (joto) [inline test/t/io/test_writer.cpp]: Extra SECTION over everything doesn't help here, probably just slows down compilation.

File: test/t/thread/test_pool.cpp [hunk 15]
```diff
@@ -48,9 +48,32 @@ TEST_CASE("number of threads in pool") {
 
 }
 
+TEST_CASE("number of threads in user-provided pool") {
+
+     SECTION("if zero requested, threads configured") {
```
Comment (joto) [inline test/t/thread/test_pool.cpp]: These should all be just TEST_CASEs without any SECTIONs. There is no common code here.

File: test/t/io/test_writer.cpp (hunk 14 above)
Comment (mloskot) [inline test/t/io/test_writer.cpp]: I decided to keep it to state clearly there are no `Pool` tests that exercise failing writer. IOW, `Pool` must be sound.
I can it remove though.

File: test/t/thread/test_pool.cpp (hunk 15 above)
Comment (mloskot) [inline test/t/thread/test_pool.cpp]: I will change it to TEST_CASEs

File: include/osmium/thread/pool.hpp (hunk 12 above)
Comment (mloskot) [inline include/osmium/thread/pool.hpp]: It buys simpler interface. unless it is desired to let user to specify the work queue too, what I have just learned.

(note, in this PR my concern is *number of threads*, other aspects of pool may deserve separate PR, who knows)

Comment (mloskot) [inline include/osmium/thread/pool.hpp]: Besides, `detail::*` used in public interface smells.

Anyhow, I'll change.

File: test/t/io/test_writer.cpp (hunk 14 above)
Comment (joto) [inline test/t/io/test_writer.cpp]: These tests are just about the `Pool` so that's okay if there is nothing failing in here, we don't need to say this explicitly.

File: include/osmium/thread/pool.hpp (hunk 12 above)
Comment (joto) [inline include/osmium/thread/pool.hpp]: Good point! How about this: make the default value `0`. Check inside the constructor if it is and then use the `detail::get_work_queue_size()`. The the documentation should say: "Set to 0 to get max queue size from OSMIUM_MAX_WORK_QUEUE_SIZE env variable". The the detail is hidden behind a good default value with the added benefit that we'll never have an actual queue size of 0 which would make no sense.

File: test/t/io/test_writer.cpp (hunk 14 above)
Comment (mloskot) [inline test/t/io/test_writer.cpp]: Agreed

**Comment by mloskot** [comment]
Second round of review requests addressed.

**Comment by mloskot** [comment]
https://travis-ci.org/osmcode/libosmium/jobs/253639146

```
 57/115 Test  #57: io_test_writer_with_mock_encoder .........***Exception: SegFault  0.04 sec
```

Hmm, only one Xcode builds is failing. I don't have access to OSX to investigate if this is actually related to this PR.

**Comment by joto** [comment]
The test failure on OSX is probably something else. The master branch also shows the same problem.

I have looked at the `Writer` constructor in more detail now and I think it is mostly fine the way you wrote it. Just initialize `m_output(nullptr)` so we have the complete list of members. And we don't need `m_pool`, just do a  `if (!options.pool) { options.pool = &thread::Pool::default_instance() }` and use `options.pool` then when creating the output.

**Comment by joto** [comment]
Now that the writer side is clear, lets look at the reader side, too. This should be even easier, because of the [struct parser_arguments](https://github.com/osmcode/libosmium/blob/master/include/osmium/io/detail/input_format.hpp#L58-L64) which can easily be extended with a `Pool` member which will then be available in all the right places. Can you add this to your pull request?

**Comment by mloskot** [comment]
@joto https://github.com/osmcode/libosmium/pull/214#issuecomment-315464056
> Just initialize m_output(nullptr) so we have the complete list of members (...)

Done.

**Comment by mloskot** [comment]
@joto https://github.com/osmcode/libosmium/pull/214#issuecomment-315465792

First stab at `Pool`-enabled `Reader` is in.

Since only `PBFParser`-based reader actually uses the pool, if there was a .pbf test data added, the `Reader` test case with user-provided `Pool` could be switched to use .pbf.

**Comment by joto** [comment]
Thanks @mloskot for the great work. I have merged the pull request now. If needed, we can do further refinement later.

**Comment by mloskot** [comment]
@joto I'm glad we could help. Thank you for your assistance.
